File: src/gtfs/ProfileAnalyzer.py
import logging
import re
from pathlib import Path

import yaml

from src.gtfs.GTFSReader import GTFSReader
from src.gtfs.models.Profile import PeriodRule, Profile
from src.gtfs.models.ServicePeriod import ServicePeriod

logger = logging.getLogger(__name__)

_DAY_INDEX = {"mon": 0, "tue": 1, "wed": 2, "thu": 3, "fri": 4, "sat": 5, "sun": 6}
_DAY_ALIASES: dict[str, set[int]] = {
    "weekday": {0, 1, 2, 3, 4},
    "weekdays": {0, 1, 2, 3, 4},
    "weekend": {5, 6},
    "weekends": {5, 6},
    "daily": set(range(7)),
    "everyday": set(range(7)),
    "all": set(range(7)),
}


class ProfileAnalyzer:
    """Build service periods from a declarative YAML profile.

    Usage as the ``period_analyzer`` of ``PipelineConverter.convert``:
        profile = ProfileAnalyzer.load("profiles/lyon.yaml")
        convert(..., period_analyzer=partial(ProfileAnalyzer.build, profile))
    """

    @staticmethod
    def load(path: str) -> Profile:
        """Load and validate a YAML profile."""
        data = yaml.safe_load(Path(path).read_text(encoding="utf-8"))
        if not isinstance(data, dict):
            raise ValueError(f"Profile '{path}' must be a YAML mapping")
        return Profile(**data)

    @staticmethod
    def _parse_days(tokens: list[str]) -> set[int]:
        result: set[int] = set()
        for token in tokens:
            t = token.strip().lower()
            if t in _DAY_ALIASES:
                result |= _DAY_ALIASES[t]
            elif "-" in t:
                start, _, end = t.partition("-")
                start, end = start.strip(), end.strip()
                if start not in _DAY_INDEX or end not in _DAY_INDEX:
                    raise ValueError(f"Invalid day range '{token}'")
                result |= set(range(_DAY_INDEX[start], _DAY_INDEX[end] + 1))
            elif t in _DAY_INDEX:
                result.add(_DAY_INDEX[t])
            else:
                raise ValueError(f"Unknown day token '{token}'")
        return result

    @staticmethod
    def _matches(rule: PeriodRule, service_id: str, active_days: set[int]) -> bool:
        if rule.service_id_matches and not re.search(rule.service_id_matches, service_id):
            return False
        if rule.days and active_days.isdisjoint(ProfileAnalyzer._parse_days(rule.days)):
            return False
        return True

    @staticmethod
    def build(profile: Profile, reader: GTFSReader) -> list[ServicePeriod]:
        """Assign every service to the profile's periods; bucket/drop the rest."""
        # service_id -> active weekday indices (empty when only in calendar_dates,
        # in which case only regex rules can match).
        services: dict[str, set[int]] = {}
        for cal in reader.calendar:
            services[cal.service_id] = {
                i
                for i, active in enumerate(
                    [cal.monday, cal.tuesday, cal.wednesday, cal.thursday,
                     cal.friday, cal.saturday, cal.sunday]
                )
                if active
            }
        for cd in reader.calendar_dates:
            services.setdefault(cd.service_id, set())

        assigned: dict[str, list[str]] = {name: [] for name in profile.periods}
        matched: set[str] = set()
        for service_id, active_days in services.items():
            for name, rule in profile.periods.items():
                if ProfileAnalyzer._matches(rule, service_id, active_days):
                    assigned[name].append(service_id)
                    matched.add(service_id)

        periods: list[ServicePeriod] = []
        for name, ids in assigned.items():
            if ids:
                periods.append(
                    ServicePeriod(
                        name=name,
                        service_ids=sorted(ids),
                        description=profile.periods[name].description
                        or f"Profile period '{name}'",
                    )
                )
            else:
                logger.warning(f"Profile period '{name}' matched no services")

        unmatched = sorted(set(services) - matched)
        if unmatched:
            if profile.unmatched == "other":
                periods.append(
                    ServicePeriod(
                        name="other",
                        service_ids=unmatched,
                        description=f"{len(unmatched)} service(s) not matched by the profile",
                    )
                )
            else:
                preview = ", ".join(unmatched[:10])
                logger.warning(
                    f"{len(unmatched)} service(s) not matched by the profile "
                    f"and dropped: {preview}{'…' if len(unmatched) > 10 else ''}"
                )

        label = profile.network or "profile"
        logger.info(f"Profile '{label}' → {len(periods)} period(s)")
        return periods

File: src/main/kotlin/com/raptor/gtfs/ProfileAnalyzer.kt
package com.raptor.gtfs

import com.charleskorn.kaml.Yaml
import com.raptor.gtfs.models.PeriodRule
import com.raptor.gtfs.models.Profile
import com.raptor.gtfs.models.ServicePeriod
import java.io.File

object ProfileAnalyzer {
    private val DAY_INDEX = mapOf(
        "mon" to 0, "tue" to 1, "wed" to 2, "thu" to 3, "fri" to 4, "sat" to 5, "sun" to 6
    )
    private val DAY_ALIASES = mapOf(
        "weekday" to setOf(0, 1, 2, 3, 4), "weekdays" to setOf(0, 1, 2, 3, 4),
        "weekend" to setOf(5, 6), "weekends" to setOf(5, 6),
        "daily" to (0..6).toSet(), "everyday" to (0..6).toSet(), "all" to (0..6).toSet()
    )

    fun load(path: String): Profile {
        val text = File(path).readText()
        return Yaml.default.decodeFromString(Profile.serializer(), text)
    }

    private fun parseDays(tokens: List<String>): Set<Int> {
        val result = mutableSetOf<Int>()
        for (token in tokens) {
            val t = token.trim().lowercase()
            when {
                DAY_ALIASES.containsKey(t) -> result.addAll(DAY_ALIASES[t]!!)
                "-" in t -> {
                    val parts = t.split("-")
                    val start = DAY_INDEX[parts[0].trim()] ?: continue
                    val end = DAY_INDEX[parts[1].trim()] ?: continue
                    result.addAll((start..end).toSet())
                }
                DAY_INDEX.containsKey(t) -> result.add(DAY_INDEX[t]!!)
            }
        }
        return result
    }

    private fun matches(rule: PeriodRule, serviceId: String, activeDays: Set<Int>): Boolean {
        if (rule.service_id_matches != null && !Regex(rule.service_id_matches).containsMatchIn(serviceId)) {
            return false
        }
        if (rule.days.isNotEmpty() && activeDays.intersect(parseDays(rule.days)).isEmpty()) {
            return false
        }
        return true
    }

    fun build(profile: Profile, reader: GTFSReader): List<ServicePeriod> {
        val services = mutableMapOf<String, Set<Int>>()
        for (cal in reader.calendar) {
            val active = mutableSetOf<Int>()
            if (cal.monday) active.add(0)
            if (cal.tuesday) active.add(1)
            if (cal.wednesday) active.add(2)
            if (cal.thursday) active.add(3)
            if (cal.friday) active.add(4)
            if (cal.saturday) active.add(5)
            if (cal.sunday) active.add(6)
            services[cal.serviceId] = active
        }
        for (cd in reader.calendarDates) {
            services.putIfAbsent(cd.serviceId, emptySet())
        }

        val assigned = profile.periods.keys.associateWith { mutableListOf<String>() }
        val matched = mutableSetOf<String>()

        for ((serviceId, activeDays) in services) {
            for ((name, rule) in profile.periods) {
                if (matches(rule, serviceId, activeDays)) {
                    assigned[name]!!.add(serviceId)
                    matched.add(serviceId)
                }
            }
        }

        val periods = mutableListOf<ServicePeriod>()
        for ((name, ids) in assigned) {
            if (ids.isNotEmpty()) {
                val desc = profile.periods[name]?.description?.takeIf { it.isNotBlank() } ?: "Profile period '$name'"
                periods.add(ServicePeriod(name, ids.sorted().toMutableList(), desc))
            }
        }

        val unmatched = services.keys.minus(matched).sorted()
        if (unmatched.isNotEmpty()) {
            if (profile.unmatched == "other") {
                periods.add(ServicePeriod("other", unmatched.toMutableList(), "${unmatched.size} service(s) not matched by the profile"))
            }
        }

        return periods
    }
}
